Fix many GCC compiler warnings and remove unused variables

JohnoKing opened the pull request:
Most of the changes in this pull request remove unused variables, functions and labels to fix `-Wunused` compiler warnings. These are the somewhat notable changes:

- src/cmd/ksh93/bltins/print.c:
  \- Removed the unused `neg` variable (patch from ksh2020: https://github.com/att/ast/pull/725).

- src/cmd/ksh93/bltins/sleep.c:
  \- Initialized `ns` to fix three `-Wsometimes-uninitialized` warnings.

- src/cmd/ksh93/include/edit.h:
  \- Increased the size of `e_search` by one byte to fix two GCC 10 `-Wstringop-truncation` warnings.

- src/cmd/ksh93/include/shell.h:
  \- The `NOT_USED` macro caused many `-Wunused-value` warnings, so it has been replaced with ksh2020's macro: https://github.com/att/ast/commit/19d0620a (to reduce code changes, the name of `NOT_USED` hasn't been changed).

- src/cmd/ksh93/sh/expand.c:
  \- Removed an unnecessary `ap = ` since `ap` is never read between `stakseek` and `stakfreeze`.

- src/cmd/ksh93/sh/nvdisc.c, src/lib/libast/regex/regsubexec.c:
  \- Fixed some indentation to silence `-Wmisleading-indentation` warnings.

JohnoKing committed:
Fix many GCC compiler warnings and remove unused variables

Most of these changes remove unused variables, functions
and labels to fix -Wunused compiler warnings.
Somewhat notable changes:

src/cmd/ksh93/bltins/print.c:
- Removed the unused 'neg' variable (patch from ksh2020: https://github.com/att/ast/pull/725).

src/cmd/ksh93/bltins/sleep.c:
- Initialized ns to fix three -Wsometimes-uninitialized warnings.

src/cmd/ksh93/include/edit.h:
- Increased the size of 'e_search' by one byte to fix two
  GCC 10 -Wstringop-truncation warnings.

src/cmd/ksh93/include/shell.h:
- The NOT_USED macro caused many -Wunused-value warnings,
  so it has been replaced with ksh2020's macro (https://github.com/att/ast/commit/19d0620a).

src/cmd/ksh93/sh/expand.c:
- Removed an unnecessary 'ap = ' since 'ap' is never read
  between stakseek and stakfreeze.

src/cmd/ksh93/sh/nvdisc.c,
src/lib/libast/regex/regsubexec.c:
- Fixed some indentation to silence -Wmisleading-indentation warnings.

hyenias commented on src/cmd/ksh93/bltins/typeset.c:
```diff
@@ -275,8 +268,8 @@ int    b_typeset(int argc,register char *argv[],Shbltin_t *context)
 				else
-				        /* n=='F' Remove possible collision with NV_UNSIGN/NV_HEXFLOAT */
-				        flag &= ~NV_HEXFLOAT;
+					/* n=='F' Remove possible collision with NV_UNSIGN/NV_HEXFLOAT */
```
I did this code. Please teach me why this is wrong?

JohnoKing replied:
You used both spaces and tabs to indent. I replaced the spaces with tabs for consistency. From policy 7 in the README:
> Your changes should match the style of the code surrounding it. Indent with **tabs**, assuming an 8-space tab width.

McDutchie replied:
Did those indentations actually produce gcc warnings? If not, it's hardly a big problem and I'm not sure I would have bothered, but it doesn't hurt to fix it either.

JohnoKing replied:
These indentations weren't the ones causing warnings (the indentations that caused warnings were in `nvdisc.c` and `regsubexec.c`).

hyenias commented on src/cmd/ksh93/bltins/typeset.c:
```diff
@@ -366,10 +359,10 @@ int    b_typeset(int argc,register char *argv[],Shbltin_t *context)
-			        {
-				        tdata.wctname = e_toupper;
-				        flag |= NV_LTOU;
+				if(!isfloat)
+				{
+					tdata.wctname = e_toupper;
+					flag |= NV_LTOU;
```
I am the author of this code. Please teach me what I did wrong.

JohnoKing replied:
Same problem with spaces and tabs as show above.

McDutchie commented on src/cmd/ksh93/include/shell.h:
```diff
@@ -37,8 +37,18 @@
+ * This is a macro that can be used to silence "unused parameter" warnings from the compiler for
+ * functions which need to accept parameters they do not use because they need to be compatible
+ * with an interface. It's similar to the Python idiom of doing `_ = param` at the top of a
+ * function in the same situation.
```
I think we could do without the Python idiom comparison in this comment.

McDutchie reviewed (changes requested):
This PR removes a lot of version ID strings of the type
```c
static const char id[] = "\n@(#)$Id: ast (AT&T Research) 2012-05-28 $\0\n";
```
Please leave all those alone. They are designed for grepping the binary using [`what`(1)](https://www.freebsd.org/cgi/man.cgi?query=what&sektion=1&manpath=FreeBSD+12.2-RELEASE+and+Ports). You can do `what arch/*/bin/ksh` to see whether a particular file is linked into the binary (not all of libast is used).  After the rest of the warnings are fixed, instead of removing these ID strings I'd be inclined to add a `#pragma GCC diagnostic ignored "-Wunused-value"` directive to `src/lib/libast/include/ast.h` where, as you can see, I've already [told clang to stop complaining so much](https://github.com/ksh93/ksh/blob/83630f9d1c10330141f69e4a891f9cf0bc8dea94/src/lib/libast/include/ast.h#L23-L25) (it is a lot more eager with warnings than gcc). Alternatively, a `NOT_USED` macro invocation could be used on those version strings, but I'm not quite sure how because those are not defined in any function.

JohnoKing committed:
Apply requested changes

JohnoKing committed:
libast: Fix more -Wmisleading-indentation warnings

JohnoKing committed:
Remove -Wunused-value pragmas

McDutchie committed:
Fix or suppress some clang warnings

Notable:

src/lib/libast/include/ast.h:
- For clang, now only suppress hundreds of -Wparentheses warnings
  as well as a few -Wstring-plus-int warnings.
  Clang's -Wparentheses warns about things like
  	if(foo = bar())
  which assigns to foo and checks the assigned value.
  Clang wants us to change this into
  	if((foo = bar()))
  Clang's -Wstring-plus-int warns about things like
  	"string"+x
  where x is an integer, e.g. "string"+3 represents the string
  "ing". Clang wants us to change that to
  	"string"[3]
  The original versions represent a prefectly valid coding style
  that was common in the 1980s and 1990s and is not going to change
  in this historic code base. (gcc does not complain about these.)

McDutchie committed:
edit/vi.c: refresh(): undef 'w' (== v) at end of function

JohnoKing committed:
Use SEARCHSIZE-1 instead of SEARCHSIZE+1
